smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -856,7 +1098,155 @@ private void writeEventBody(GenerationContext context, StructureShape event) {
+            readHost(context, operation);
+            readRequestHeaders(context, operation, bindingIndex, "output");
+            List<HttpBinding> documentBindings = readRequestBody(context, operation, bindingIndex);
+            // Track all shapes bound to the document so their deserializers may be generated.
+            documentBindings.forEach(binding -> {
+                Shape target = model.expectShape(binding.getMember().getTarget());
+                deserializingDocumentShapes.add(target);
+            });
+            writer.write("return Promise.resolve(contents);");
+        });
+        writer.write("");
+    }
+
+    private void readQueryString(
+            GenerationContext context,
+            OperationShape operation,
+            HttpBindingIndex bindingIndex
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        List<HttpBinding> queryBindings = bindingIndex.getRequestBindings(operation, Location.QUERY);
+        if (queryBindings.isEmpty()) {
+            return;
+        }
+        writer.write("const query = output.query");
+        writer.openBlock("if (query !== undefined && query !== null) {", "}", () -> {
+            for (HttpBinding binding : queryBindings) {
+                String memberName = context.getSymbolProvider().toMemberName(binding.getMember());
+                writer.openBlock("if (query['$L'] !== undefined) {", "}", binding.getLocationName(), () -> {
+                    Shape target = context.getModel().expectShape(binding.getMember().getTarget());
+                    String expectedType = "string";
+                    if (target instanceof CollectionShape) {
+                        expectedType = "string[]";
+                    }
+                    String headerValue = getOutputValue(context, binding.getLocation(),
+                            "query['" + binding.getLocationName() + "'] as " + expectedType,
+                            binding.getMember(), target);
+                    writer.write("contents.$L = $L;", memberName, headerValue);
+                });
+            }
+        });
+    }
+
+    private void readPath(
+            GenerationContext context,
+            OperationShape operation,
+            HttpBindingIndex bindingIndex,
+            HttpTrait trait
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        List<HttpBinding> pathBindings = bindingIndex.getRequestBindings(operation, Location.LABEL);
+        if (pathBindings.isEmpty()) {
+            return;
+        }
+        StringBuilder pathRegexBuilder = new StringBuilder();
+        for (Segment segment : trait.getUri().getSegments()) {
+            pathRegexBuilder.append("/");
+            if (segment.isLabel()) {
+                // Create a named capture group for the segment so we can grab it later without regard to order.
+                pathRegexBuilder.append(String.format("(?<%s>.*)", segment.getContent()));
```

Comment:
In ... other implementations ... this uses `[^/]+` for non-greedy segments and `.+` for greedy ones. There may be cases with path patterns where this regex causes non-greedy segments to span `/`s:

`/(?<label>.*)/(?<greedy>.*)` would probably not match `/foo/bar/baz` in the way you want.

Reply:
Nice catch, thanks!